Create protocol-spec for all fork choice patterns - Closes #4129

nazarhussain opened the pull request:
### What was the problem?

There were no JSON based protocol specs for Fork Choices. 

### How did I solve it?

Created Specs of Specs generator for Fork Choice and used it in unit tests. 

### How to manually test it?

Run jest unit tests. 

### Review checklist

- [ ] The PR resolves #4129
- [ ] All new code is covered with unit tests
- [ ] Relevant integration / functional tests are added
- [ ] Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
- [ ] Documentation has been added/updated

shuse2 commented on protocol-specs/package.json:
```diff
@@ -11,22 +11,28 @@
+		"@liskhq/lisk-validator": "0.2.1",
+		"lodash": "4.17.15",
+		"randomstring": "^1.1.5",
+		"sodium-native": "^2.4.6"
```
dependencies should be `exact`

shuse2 commented on protocol-specs/utils/bft/index.js:
```diff
@@ -68,8 +70,53 @@ const generateBlockHeadersSeries = ({ activeDelegates, count }) => {
+				.generate({ charset: '0123456789ABCDE', length: 64 })
+				.toLowerCase(),
+		height,
+		previousBlockId: previousBlockId || null,
```
prefer not to use `null`

shuse2 commented on protocol-specs/generators/bft_processing/index.js:
```diff
@@ -313,10 +314,202 @@ const bftInvalidBlockHeaderTestSuiteGenerator = ({
+					'DISCARD: Received invalid block for current state of chain',
+				initialState,
+				input: {
+					// Any block with lower height than last block is invalid
```
this comment comment can be improved because if the `maxHeightPrevoted` is higher but height is lower, then it will be `different chain`

nazarhussain commented:
@pablitovicente I don't agree that generators should always generate same output on each run. Rather they should generate valid required output on reach run. 

And its the same as already for other generators as well. Try it out: 

```
cd protocol-specs && npm run generate-all-specs
```

And you will notice following files changed: 

```
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   generator_outputs/block_processing_votes/invalid_block_processing_vote_all_delegates_in_one_transaction.json
	modified:   generator_outputs/block_processing_votes/invalid_block_processing_vote_already_voted_delegate.json
	modified:   generator_outputs/block_processing_votes/invalid_block_processing_vote_no_delegate.json
	modified:   generator_outputs/block_processing_votes/valid_block_processing_vote_all_delegates.json
```

pablitovicente commented:
> @pablitovicente @shu I don't agree that generators should always generate same output on each run. Rather they should generate valid payload or JSON output on reach run.
> 
> And its the same as already for other generators as well. Try it out:
> 
> ```
> cd protocol-specs && npm run generate-all-specs
> ```
> 
> And you will notice following files changed:
> 
> ```
> Changes not staged for commit:
>   (use "git add <file>..." to update what will be committed)
>   (use "git checkout -- <file>..." to discard changes in working directory)
> 
> 	modified:   generator_outputs/block_processing_votes/invalid_block_processing_vote_all_delegates_in_one_transaction.json
> 	modified:   generator_outputs/block_processing_votes/invalid_block_processing_vote_already_voted_delegate.json
> 	modified:   generator_outputs/block_processing_votes/invalid_block_processing_vote_no_delegate.json
> 	modified:   generator_outputs/block_processing_votes/valid_block_processing_vote_all_delegates.json
> ```

Then those will also to eventually be fixed. It's a bug that they're generating different files.

nazarhussain commented:
@pablitovicente @shuse2 Protocol specs are generators for valid and invalid cases of the protocol, these are not the some kind of static fixtures. So I tend not to agree that these should always generate same output. Till we protocol specific valid payload on each run we should be good to go.

shuse2 commented:
@nazarhussain 
We had the discussion last year, and for now we decided to output the consistent json file. Until we find the good reason to make it random, let's keep it as static

pablitovicente commented:
> @pablitovicente @shuse2 Protocol specs are generators for valid and invalid cases of the protocol, these are not the some kind of static fixtures. So I tend not to agree that these should always generate same output. Till we protocol specific valid payload on each run we should be good to go.

Well for the sake of consistency I think all output should be the same on each run (the case you quote on your first answer to this is because someone in another PR forgot to commit the JSON output but the output is constant)

nazarhussain commented:
@shuse2 @pablitovicente Please review the PR again.

nazarhussain commented:
@pablitovicente This particular spec is not static. 

```
generator_outputs/block_processing_votes/invalid_block_processing_vote_no_delegate.json
```

Every time you run you get different output. I will look into that bug and add updated JSONs with the issue #4672
